### ISSUE TITLE
Make one folder for all screenshots / screenshot browser

### ISSUE THREAD

**Issue by SuperFredi** [PrismLauncher/PrismLauncher#1027]
### Role

I have many different instances that I use regularly and I take many screenshots.

### Suggestion

I want a central screenshot folder, where all screenshots of each instance are stored.

### Benefit

Searching every instance for "that one screenshot" is tedious and time consuming. 

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

By having a central screenshots folder it is easier to browse all of your screenshots. You also don't need to worry that when you delete an instance, that you forgot to backup your screenshots.

If that is not possible because Minecraft's programming prevents it, an alternative could be a 'browse screenshots' button / window, where you can browse all of your screenshots from all of your instances, in the top bar. 

Something like this, for example:
![grafik](https://user-images.githubusercontent.com/93713786/233667593-7ffa35fe-c4fc-46bf-8443-7b0a236527fe.png)

**Comment by maxashen** [comment]
The screenshot directory appears to be hardcoded by the _Minecraft: Java Edition_ executable relative to the working directory (.minecraft directory). However, in theory you could work around this limitation by using Prism Launcher's existing custom commands feature. You'll have to write a short program that takes the absolute path of the instance as a command line option and moves all screenshots to the desired folder. Then simply have Prism call that program whenever you exit the game, from Settings->Custom Commands->Post-exit commands.

I successfully tested this work around on Windows 10 with the following program (Java; you could write a shell script instead):

```
package moveall;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.FileVisitResult;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.attribute.BasicFileAttributes;
import static java.nio.file.FileVisitResult.CONTINUE;
import static java.nio.file.StandardCopyOption.*;

public class MoveAll
{
	public static class MoveAllFileVisitor extends SimpleFileVisitor<Path>
	{
		private final Path dst;
		
		public MoveAllFileVisitor(Path path)
		{
			this.dst = path;
		}
		
		@Override
		public FileVisitResult visitFile(Path file, BasicFileAttributes attr)
		{
			if(attr.isRegularFile())
			{
				try
				{
					Files.move
					(
						file,
						this.dst.resolve(file.getFileName()),
						REPLACE_EXISTING
					);
				}
				catch(IOException e)
				{
				}
			}
			
			return CONTINUE;
		}
	}
	
	public static void main(String[] args)
	{
		if(args.length == 2)
		{
			File dst = new File(args[1]);
			dst.mkdirs();
			
			MoveAllFileVisitor fv = new MoveAllFileVisitor(dst.toPath());
			try
			{
				Files.walkFileTree(Paths.get(args[0]), fv);
			}
			catch(IOException e)
			{
				e.printStackTrace();
			}
		}
	}
}
```

I compiled with Java 17, but it should work with Java 7+ or whenever NIO was added.

`javac -d out\ MoveAll.java`
`jar cfe MoveAll.jar moveall.MoveAll -C out\ moveall`

And I used this option in the Prism Launcher, Settings->Custom Commands->Post-exit:

`java -jar "C:/path/to/MoveAll.jar" "$INST_MC_DIR/screenshots/" "C:/path/to/pictures/folder/$INST_NAME/"`

**Comment by DeveloperBlue** [comment]
Would it be possible to have Prism create a symlink/junction to a folder? And we can set a global default folder directory in settings?

**Comment by Aces-and-Jacks** [comment]
I'm gonna express @Hexasan 's idea from #2080 but I think it's a great idea and would help with instance deleting. 

> When deleting an old instance, It should ask you if you want to save screenshots in the instance, if checked the screenshots would be moved to a seperate folder(possibly selected by the user).


Then from what I added on to it. 

> Ohh, like maybe have a central screenshots folder (similar to the central mods folder) and move screenshots to there, if a location hasn't been set by the user. Maybe even generate another folder in that folder per instance name, if possible.

I know it's similar to what @SuperFredi said but I'm mainly just adding onto it

**Comment by Riva3000** [comment]
> Would it be possible to have Prism create a symlink/junction to a folder? And we can set a global default folder directory in settings?

I can confirm that at least Minecraft ver 1.20.1 and 1.20.4 are able to save screenshot files into manually created junction named "screenshots" that replaces the original screenshots folder in .minecraft root dir.
So that can be a route forward, at least for Windows / NTFS.

**Comment by electricsteve** [comment]
You could also implement a screenshot browser in a way where all screenshots are viewable but not moved to a central folder. Like select the instances you want to see the screenshots from, execute mass actions on them, like export and delete.
A setting could also be added to control how they are saved, in the way mentioned above or a way mentioned earlier in this conversation.

**Comment by Deskilling** [comment]
Great idea! (hopefully not dead)

**Comment by ninetailedtori** [comment]
The symlink/hardlink solution works fine on Linux as well, and probably should be the default solution, only falling back on using the initial folder as a temp before on instance closing, moving all images over to the main Prism-settings set folder if the fs doesn't support it. I don't see why that can't be an option within Prism's settings, it seems within capabilities of the launcher, so hopefully we can get this. It would be cool to be able to automate it rather than having to make symlinks manually.

**Comment by Redo11** [comment]
Yes, please! I would love a gallery tab being added, so I can sort through screenshots across multiple instances. Also, duplicate recognition will be a must, since screenshots are often copied when cloning an instance.

**Comment by Redo11** [comment]
Also I don't believe the approach of symlinking the images into a singular folder is a good one. Not all filesystems support that (looking at you EXFAT), and such filesystems are common on removable media, such as pendrives.
It's better if the gallery can just browse screenshots from all instances, and if a player wants to, there could be an option to copy or move image to a special global prism screenshot folder/os-defined one.
There could be a switch in settings that "auto moves" or "auto copies" all screenshot into the global folder. Such thing would obviously only work when prism is open/running in the background and users should be informed about it. It is also important to check whenever the folders are on the same drive and partition, as moving operations between differing filesystems will cause additional disk usage and while it would be similar to copying them, doing so in background, while Minecraft is playing, could cause lag, especially with mods that allow for video recording or high-resolution/quality screenshots (fabrishot). Also it wastes drive cycles and space on COW filesystems like BTRFS.
I also would expect it to work similarly to the existing per-instance gallery with movable icons, selection mode and basic actions.
An important note is, that screenshot-managing mods will be incompatible with such approach, since they do not expect them to be moved, but it's up to the user to define their workflow.

**Comment by Redo11** [comment]
Other option is to symlink each and every screenshot folder in instances to a global one, but as I mentioned, it comes with  limitations on certain filesystems. _(Microsoft, can you finally support any good filesystem?! It's time already! Even ReactOS runs on BTRFS!)_

**Comment by ninetailedtori** [comment]
I think that having the option appear if both the minecraft and screenshot destination drive are link-supported, is more than ok. But having a backup solution of shared folder systems through copy-moving, similar to Steam's screenshot manager, can work wonders for that as well. CoW systems such as btrfs and bcachefs, we can actively implement an optimised CoW shared folder solution anyway. btrfs and other CoW systems sport a reflink support, and many other solutions that work around multiple inode linkage, and would work better for there. Implementing a fs checker is not too difficult in any situation, but if the fs types change, we will need to implement proper data handling on that. Or a warning that fs changing is not officially supported especially if linking across volumes. LVMs may need to be handled specially as well.

**Comment by OffsetMonkey538** [comment]
I'd love if I could have screenshot directories of instances symlinked to something like `$XDG_PICTURES_DIR/Screenshots/Minecraft/$INSTANCE_NAME`

**Comment by electricsteve** [comment]
> I'd love if I could have screenshot directories of instances symlinked to something like `$XDG_PICTURES_DIR/Screenshots/Minecraft/$INSTANCE_NAME`

Agree, but the problem is windows. It needs to be compatible with both windows, linux and macos.

**Comment by ninetailedtori** [comment]
windows just *has* to suck, doesn't it ;w;

**Comment by ZepsiZola** [comment]
Was about to make an issue suggesting this exact feature. So +1 on this. I think the symlink idea is the best. Just make the option toggleable (off by default) and not give the option for Windows.

Edit: Also didn't expect *everyone* here to be a linux user lmao

**Comment by OffsetMonkey538** [comment]
Does ntfs not have symlinks or something? Why wouldn't using them be compatible with windows?
Or was @electricsteve commenting on using the xdg variable?

**Comment by Riva3000** [comment]
NTFS does have symbolic links. Many different flavors of them actually.

- It has the 'classic' symbolic links, for files and directories, that function the same / similar to other file systems. They also supports network (LAN) paths (via UNC paths).
- It also has 'junctions' for dirs that behave slightly differently. 

All the above are implemented via NTFS 'Reparse points'.

- It also has the 'classic' hardlinks for files. Those are recorded natively inside Master File Table.

Prism launcher already supports and recognizes NTFS symbolic links. I tested. I use them a lot with PL on Windows.
Minecraft itself supports symbolic links (but limits them for security reasons). It also either supports or is oblivious to junctions = treats them as regular dirs and uses them without a problem.

With all respect to @Redo11, I am not sure why he brings up arguably exotic corner cases like EXFAT and BTRFS as supposed dealbreakers ?

I am not pushing for symlinks solution. Just wanted to give some information and clear possible misconceptions.

[Reparse points - Microsoft Learn](https://learn.microsoft.com/en-us/windows/win32/fileio/reparse-points)
[Creating Symbolic Links - Microsoft Learn](https://learn.microsoft.com/en-us/windows/win32/fileio/creating-symbolic-links)
[Hard links and junctions - Microsoft Learn](https://learn.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions)

**Comment by electricsteve** [comment]
> Does ntfs not have symlinks or something? Why wouldn't using them be compatible with windows? Or was [@electricsteve](https://github.com/electricsteve) commenting on using the xdg variable?

I think I have fallen victim to misinformation somewhere on the internet. I thought that windows didn't support symlinks. But from what I read from @Riva3000 it seems I have been proven wrong.

**Comment by Redo11** [comment]
> With all respect to [@Redo11](https://github.com/Redo11), I am not sure why he brings up arguably exotic corner cases like EXFAT and BTRFS as supposed dealbreakers ?

Im talking about specifically EXFAT, since it is often used on removable media. BTRFS is just an example, that a Windows recreation project(ReactOS) supports better filesystems.

**Comment by cultab** [comment]
I made a script for this and posted it in a discussion https://github.com/orgs/PrismLauncher/discussions/5094

**Comment by Ratstail91** [comment]
@cultab that mod doesn't seem to work, at least in 26.2.

This would be a great feature, though given the issues, another solution would be symlinks from the desired destination (such as the photos folder) to each installed MC version, possibly as a post-install script. It's not perfect, but would allow uploading to the browser via a dialog box much easier.
